File: feedback_service/feedback_service/repositories/interfaces/feedback.py
"""Feedback repository interfaces."""

from abc import abstractmethod
from collections.abc import Sequence
from datetime import datetime

from feedback_service.models import Comment, ExperienceRating, PulseSurvey

from .base import BaseRepository


class IPulseSurveyRepository(BaseRepository[PulseSurvey, int]):
    """Interface for PulseSurvey repository."""

    @abstractmethod
    async def find_by_user(self, user_id: int, skip: int = 0, limit: int = 100) -> Sequence[PulseSurvey]:
        """Find pulse surveys by user ID."""
        raise NotImplementedError

    @abstractmethod
    async def get_by_user(
        self,
        user_id: int | None,
        from_date: datetime | None,
        to_date: datetime | None,
        skip: int,
        limit: int,
    ) -> tuple[Sequence[PulseSurvey], int]:
        """Get pulse surveys with optional user filter and date range."""
        raise NotImplementedError

    @abstractmethod
    async def get_stats(
        self,
        user_id: int | None,
        from_date: datetime | None,
        to_date: datetime | None,
    ) -> dict:
        """Get aggregate statistics for pulse surveys."""
        raise NotImplementedError

    @abstractmethod
    async def get_rating_distribution(
        self,
        user_id: int | None,
        from_date: datetime | None,
        to_date: datetime | None,
    ) -> dict[int, int]:
        """Get count of each rating (1-10)."""
        raise NotImplementedError

    @abstractmethod
    async def get_anonymity_stats(
        self,
        department_id: int | None,
        from_date: datetime | None,
        to_date: datetime | None,
    ) -> dict:
        """Get stats for anonymous vs attributed feedback."""
        raise NotImplementedError


class IExperienceRatingRepository(BaseRepository[ExperienceRating, int]):
    """Interface for ExperienceRating repository."""

    @abstractmethod
    async def find_by_user(self, user_id: int, skip: int = 0, limit: int = 100) -> Sequence[ExperienceRating]:
        """Find experience ratings by user ID."""
        raise NotImplementedError

    @abstractmethod
    async def get_by_user(
        self,
        user_id: int | None,
        from_date: datetime | None,
        to_date: datetime | None,
        min_rating: int | None,
        max_rating: int | None,
        skip: int,
        limit: int,
    ) -> tuple[Sequence[ExperienceRating], int]:
        """Get experience ratings with optional filters."""
        raise NotImplementedError

    @abstractmethod
    async def get_stats(
        self,
        user_id: int | None,
        from_date: datetime | None,
        to_date: datetime | None,
    ) -> dict:
        """Get aggregate statistics for experience ratings."""
        raise NotImplementedError

    @abstractmethod
    async def get_rating_distribution(
        self,
        user_id: int | None,
        from_date: datetime | None,
        to_date: datetime | None,
    ) -> dict[int, int]:
        """Get count of each rating (1-5)."""
        raise NotImplementedError

    @abstractmethod
    async def get_anonymity_stats(
        self,
        department_id: int | None,
        from_date: datetime | None,
        to_date: datetime | None,
    ) -> dict:
        """Get stats for anonymous vs attributed feedback."""
        raise NotImplementedError


class ICommentRepository(BaseRepository[Comment, int]):
    """Interface for Comment repository."""

    @abstractmethod
    async def find_by_user(self, user_id: int, skip: int = 0, limit: int = 100) -> Sequence[Comment]:
        """Find comments by user ID."""
        raise NotImplementedError

    @abstractmethod
    async def get_by_user(
        self,
        user_id: int | None,
        from_date: datetime | None,
        to_date: datetime | None,
        search: str | None,
        has_reply: bool | None,
        skip: int,
        limit: int,
    ) -> tuple[Sequence[Comment], int]:
        """Get comments with optional filters."""
        raise NotImplementedError

    @abstractmethod
    async def add_reply(
        self,
        comment_id: int,
        reply: str,
        replied_by: int,
    ) -> Comment | None:
        """Add a reply to a comment."""
        raise NotImplementedError

    @abstractmethod
    async def get_anonymity_stats(
        self,
        department_id: int | None,
        from_date: datetime | None,
        to_date: datetime | None,
    ) -> dict:
        """Get stats for anonymous vs attributed comments."""
        raise NotImplementedError

    @abstractmethod
    async def get_reply_eligible_comments(
        self,
        department_id: int | None,
        from_date: datetime | None,
        to_date: datetime | None,
        skip: int,
        limit: int,
    ) -> tuple[Sequence[Comment], int]:
        """Get comments that can be replied to (non-anonymous or anonymous with contact)."""
        raise NotImplementedError

File: admin_web/src/shared/types/feedback.ts
/** Feedback types for admin web. */

export interface PulseSurvey {
  id: number;
  user_id: number | null;
  is_anonymous: boolean;
  rating: number;
  submitted_at: string;
  department_id: number | null;
  position_level: string | null;
}

export interface ExperienceRating {
  id: number;
  user_id: number | null;
  is_anonymous: boolean;
  rating: number;
  submitted_at: string;
  department_id: number | null;
}

export interface Comment {
  id: number;
  user_id: number | null;
  is_anonymous: boolean;
  comment: string;
  submitted_at: string;
  department_id: number | null;
  allow_contact: boolean;
  contact_email: string | null;
  reply: string | null;
  replied_at: string | null;
  replied_by: number | null;
}

export interface FeedbackListResponse<T> {
  items: T[];
  total: number;
  skip: number;
  limit: number;
}

export interface PulseStats {
  average_rating: number | null;
  total_responses: number;
  min_rating: number | null;
  max_rating: number | null;
  rating_distribution: Record<number, number>;
}

export interface ExperienceStats {
  average_rating: number | null;
  total_ratings: number;
  min_rating: number | null;
  max_rating: number | null;
  rating_distribution: Record<number, number>;
}

export interface AnonymityStats {
  anonymous: {
    average_rating?: number | null;
    count: number;
    with_contact?: number;
  };
  attributed: {
    average_rating?: number | null;
    count: number;
  };
}

export type FeedbackType = "pulse" | "experience" | "comment";

export interface FeedbackItem {
  id: number;
  type: FeedbackType;
  user_id: number | null;
  is_anonymous: boolean;
  rating?: number;
  comment?: string;
  submitted_at: string;
  department_id: number | null;
  allow_contact?: boolean;
  contact_email?: string | null;
  reply?: string | null;
  replied_at?: string | null;
  replied_by?: number | null;
}
